transports/quic/tests/smoke.rs

```diff
@@ -183,6 +183,98 @@ fn concurrent_connections_and_streams_tokio() {
         .quickcheck(prop::<quic::tokio::Provider> as fn(_, _) -> _);
 }
 
+#[cfg(feature = "tokio")]
+#[tokio::test]
+async fn draft_29_support() {
+    use std::task::Poll;
+
+    use futures::{future::poll_fn, select};
+    use libp2p::TransportError;
+
+    let _ = env_logger::try_init();
+
+    let (_, mut a_transport) =
+        create_transport::<quic::tokio::Provider>(|cfg| cfg.support_draft_29 = true);
+    let (_, mut b_transport) =
+        create_transport::<quic::tokio::Provider>(|cfg| cfg.support_draft_29 = true);
+
+    // If a server supports draft-29 all its QUIC addresses can be dialed on draft-29 or version-1
+    let a_quic_addr = start_listening(&mut a_transport, "/ip4/127.0.0.1/udp/0/quic").await;
+    let a_quic_mapped_addr = a_quic_addr
+        .clone()
+        .into_iter()
+        .map(|p| {
+            if matches!(p, Protocol::QuicV1) {
+                Protocol::Quic
+            } else {
+                p
+            }
+        })
+        .collect();
+    let a_quicv1_addr = start_listening(&mut a_transport, "/ip4/127.0.0.1/udp/0/quic-v1").await;
+    let a_quic_v1_mapped_addr = a_quic_addr
+        .clone()
+        .into_iter()
+        .map(|p| {
+            if matches!(p, Protocol::Quic) {
+                Protocol::QuicV1
+            } else {
+                p
+            }
+        })
+        .collect();
+
+    connect(&mut a_transport, &mut b_transport, a_quic_addr.clone()).await;
+    connect(&mut a_transport, &mut b_transport, a_quic_mapped_addr).await;
+    connect(&mut a_transport, &mut b_transport, a_quicv1_addr).await;
+    connect(&mut a_transport, &mut b_transport, a_quic_v1_mapped_addr).await;
+
+    let (_, mut c_transport) =
+        create_transport::<quic::tokio::Provider>(|cfg| cfg.support_draft_29 = false);
+    assert!(matches!(
+        c_transport.dial(a_quic_addr.clone()),
+        Err(TransportError::MultiaddrNotSupported(_))
+    ));
+
+    // Test disabling draft-29 on a server.
+    let (_, mut d_transport) =
+        create_transport::<quic::tokio::Provider>(|cfg| cfg.support_draft_29 = false);
+    assert!(matches!(
+        d_transport.listen_on("/ip4/127.0.0.1/udp/0/quic".parse().unwrap()),
+        Err(TransportError::MultiaddrNotSupported(_))
+    ));
+    let d_quic_v1_addr = start_listening(&mut a_transport, "/ip4/127.0.0.1/udp/0/quic").await;
+    let d_quic_addr_mapped = d_quic_v1_addr
+        .into_iter()
+        .map(|p| {
+            if matches!(p, Protocol::QuicV1) {
+                Protocol::Quic
+            } else {
+                p
+            }
+        })
+        .collect();
+    let dial = b_transport.dial(d_quic_addr_mapped).unwrap();
+    let drive_transports = poll_fn::<(), _>(|cx| {
+        let _ = b_transport.poll_next_unpin(cx);
+        let _ = d_transport.poll_next_unpin(cx);
+        Poll::Pending
+    });
+    select! {
```

Comment:
@mxinden converted me to prefer the `select` fn. Can we convert you too? :)

Reply:
Can you point me to the discussion about the reasoning to preferring the fn? Right now I have a slight preference for the macro because of the readability.

Reply:
Basically the rationale is:

- Plays better with IDEs
- Much easier to understand
- Clear type-safety in how to handle the various outcomes
- No fuse needed
- No magic syntax to learn

I don't know where exactly we had this discussion, so can't link you to it unfortunately.